File: backend/app/models/auto_apply.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.models.common import ApiModel

AutoApplyStatusLiteral = Literal[
    "queued", "awaiting_approval", "tailoring", "submitted", "skipped"
]


class AutoApplyQueueItemOutput(ApiModel):
    id: str
    profile_id: str
    job_posting_id: str | None
    role: str
    company: str
    location: str | None
    match_score: float | None
    status: AutoApplyStatusLiteral
    note: str | None
    created_at: datetime


class AutoApplyQueueStats(ApiModel):
    in_queue: int
    applied_today: int
    awaiting_approval: int
    avg_match: int


class AutoApplyQueueListResponse(ApiModel):
    items: list[AutoApplyQueueItemOutput]
    total: int
    stats: AutoApplyQueueStats


class AutoApplyEnqueueRequest(ApiModel):
    profile_id: str
    job_posting_id: str
    note: str | None = None


class AutoApplyUpdateRequest(ApiModel):
    status: AutoApplyStatusLiteral

File: webapp/src/schemas/auto-apply.ts
import { z } from 'zod';

export const autoApplyStatusSchema = z.enum([
  'queued', 'awaiting_approval', 'tailoring', 'submitted', 'skipped',
]);
export type AutoApplyStatus = z.infer<typeof autoApplyStatusSchema>;

export const autoApplyQueueItemSchema = z.object({
  id: z.string(),
  profile_id: z.string(),
  job_posting_id: z.string().nullable(),
  role: z.string(),
  company: z.string(),
  location: z.string().nullable(),
  match_score: z.number().nullable(),
  status: autoApplyStatusSchema,
  note: z.string().nullable(),
  created_at: z.string(),
});

export const autoApplyQueueStatsSchema = z.object({
  in_queue: z.number(),
  applied_today: z.number(),
  awaiting_approval: z.number(),
  avg_match: z.number(),
});

export const autoApplyQueueListSchema = z.object({
  items: z.array(autoApplyQueueItemSchema),
  total: z.number(),
  stats: autoApplyQueueStatsSchema,
});

export type AutoApplyQueueItem = z.infer<typeof autoApplyQueueItemSchema>;
export type AutoApplyQueueList = z.infer<typeof autoApplyQueueListSchema>;
